Implement host->guest postMessage to update theme details

sfc-gh-tteixeira opened the pull request:
## 📚 Context

When iframing Streamlit, this PR makes it possible for the host frame to update Streamlit's theme in some detail.

Note that this is not meant to be a public API (yet??) 

What this PR does
- Adds frontend/hostframe.html to test host-guest communication. See that file for usage.
  - Allows localhost postmessage when in dev mode, so hostframe.html can do its thing.
- Adds ability to theme certain properties using postMessage from Host to Guest. The themable
  properties are:
  - All normal theme colors from config.toml
  - Widget background color. This was separated from secondaryBackgroundColor, which is now only
    used in the sidebar.
  - Widget border radius. NOTE: This is not working quite right! Need to figure out why before merging.
    - To support this, we changed checkboxes so they use radii.sm. This fixes the checkbox when using
      Snowsight's theme, and is a barely noticeable change when using Streamlit's.
    - Now only md and sm radii are used in streamlit, and sm only used in st.checkbox. So these
      are called basewidgetradius and checkboxradius in the postmessage.
  - Font sizes. You can adjust md, sm, and twosm font sizes, but not anything else. This is because (1) the first two are ubiquitous, (2) the last one is so small that it needs to be adjusted on a per-pixel basis, and (3) everything else is derived from md, through the use of rem units, so they auto-adjust. We can always support specifically setting those in the future, though.
  - Font families. You can pass @font-face declarations to Streamlit via postMessage now.


What's missing:
* Tests for the new functionality
* Fixing the widget radius bug mentioned above. 


---

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

_Insert screenshot of your updated UI/code here_

**Current:**

_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [ ] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #XXXX

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-tteixeira committed:
Implement host->guest postMessage that customizes the theme.

sfc-gh-tteixeira committed:
Pipe new theme primitives into emotion theme.

sfc-gh-tteixeira committed:
Update baseweb theme.

sfc-gh-tteixeira committed:
Pipe font sizes into globalStyles. Replace absolute sizes with px values (since theme uses px) and keep relative sizes (rem) untouched.

sfc-gh-tteixeira committed:
Adjust checkbox styles. Made checkbox squarer in the process.

sfc-gh-tteixeira committed:
Fix numberinput styles. This component is built by us, not directly by baseweb.

sfc-gh-tteixeira committed:
Propagate new styles to sidebar elements.

sfc-gh-tteixeira committed:
Add ability to try out host->guest communication locally.

sfc-gh-tteixeira committed:
Fix tests.

sfc-gh-tteixeira committed:
Add tests.

mayagbarnes commented on frontend/src/hocs/withHostCommunication/types.ts:
```diff
@@ -108,6 +108,10 @@ export type IHostToGuestMessage = {
     }
+  | {
+      type: "SET_CUSTOM_THEME_CONFIG"
+      theme: ICustomThemeConfig
```
Nit: Would prefer that the `theme` field here be renamed to `themeInfo` consistent with the "SET_THEME_CONFIG" message type

sfc-gh-tteixeira replied:
I just took a second look, and the problem is that the types aren't the same. One is an `ICustomThemeConfig` and the other is an `ExportedTheme`. Wouldn't that be confusing?

Here are the two types:
```
  | {
      type: "SET_CUSTOM_THEME_CONFIG"
      themeInfo: ICustomThemeConfig
    }
```

vs

```
  | {
      type: "SET_THEME_CONFIG"
      themeInfo: ExportedTheme
    }
```

sfc-gh-tteixeira replied:
Also related: should I add the new fields to `ExportedTheme` as well, just for uniformity?

mayagbarnes replied:
IMO not confusing as they are both passing theme information - elements of theme as opposed to a complete theme - despite different types, but comfortable wherever you settle. 
Also I would vote no on adding extraneous fields to ExportedTheme simply for uniformity - though should be examined as part of the theming refactor.

mayagbarnes commented:
Just in case its helpful, worked on a couple potential FE tests ->  **For `ThemedApp.test.tsx`**
```javascript
  import { act } from "react-dom/test-utils"
  import FontFaceDeclaration from "./components/core/FontFaceDeclaration"

  const mockCustomThemeConfig = {
    primaryColor: "#1A6CE7",
    backgroundColor: "#FFFFFF",
    secondaryBackgroundColor: "#F5F5F5",
    textColor: "#1A1D21",
    widgetBackgroundColor: "#FFFFFF",
    widgetBorderColor: "#D3DAE8",
    fontFaces: [
      {
        family: "Inter",
        url: "https://rsms.me/inter/font-files/Inter-Regular.woff2?v=3.19",
        weight: 400,
      },
    ],
  }


  it("handles custom theme sent from Host", () => {
    const wrapper = mount(<ThemedApp />)
    
    let fontFaceComponent = wrapper.find(FontFaceDeclaration)
    expect(fontFaceComponent.exists()).toBe(false)

    const props = wrapper.find(AppWithScreencast).props()
    act(() => {
      props.theme.setImportedTheme(mockCustomThemeConfig)
    })
    
    wrapper.update()

    const updatedTheme: ThemeConfig = wrapper.find(AppWithScreencast).props().theme.activeTheme
    expect(updatedTheme.name).toBe(CUSTOM_THEME_NAME)
    expect(updatedTheme.emotion.genericColors.primary).toBe(mockCustomThemeConfig.primaryColor)

    fontFaceComponent = wrapper.find(FontFaceDeclaration)
    expect(fontFaceComponent.exists()).toBe(true)
    expect(fontFaceComponent.props().fontFaces).toEqual(mockCustomThemeConfig.fontFaces)
  })
```

**For `withHostCommunication.test.tsx`**
```javascript
const mockCustomThemeConfig = {
  primaryColor: "#1A6CE7",
  backgroundColor: "#FFFFFF",
  secondaryBackgroundColor: "#F5F5F5",
  textColor: "#1A1D21",
  widgetBackgroundColor: "#FFFFFF",
  widgetBorderColor: "#D3DAE8",
}

  it("can process a received SET_CUSTOM_THEME_CONFIG message", () => {
    act(() => {
      dispatchEvent(
        "message",
        new MessageEvent("message", {
          data: {
            stCommVersion: HOST_COMM_VERSION,
            type: "SET_CUSTOM_THEME_CONFIG",
            theme: mockCustomThemeConfig,
          },
          origin: "http://devel.streamlit.test",
        })
      )
    })

    wrapper.update()

    const theme = wrapper.find(TestComponent).prop("theme")
    expect(theme.setImportedTheme).toHaveBeenCalledWith(mockCustomThemeConfig)
  })
```

sfc-gh-tteixeira committed:
Fix widgets that got inadvertently changed before.

sfc-gh-tteixeira committed:
No-op. Stop using px values in globalStyles when not necessary.

sfc-gh-tteixeira committed:
Rename themeConfigObj -> themeInfo

sfc-gh-tteixeira committed:
Improve hostframe.html instructions

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/sis-updates' into theme-postmessage

sfc-gh-jgarcia committed:
Fix return type for addPxUnit function

sfc-gh-jgarcia committed:
Remove unused import

sfc-gh-jgarcia committed:
Fix duplicated import

sfc-gh-jgarcia committed:
Make react in scope and add return type

sfc-gh-jgarcia committed:
Update ThemedApp and withHostCommunication tests

sfc-gh-tteixeira committed:
Fix theme value propagation bug.

This addresses the radius propagation issue, and allows us to remove
hacky code that manually propagated the widgetBackgroundColor and
widgetBorderColor in ThemedSidebar.tsx.

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/sis-updates' into theme-postmessage

sfc-gh-tteixeira committed:
Wrap ThemedApp callbacks in useCallback().

sfc-gh-jgarcia committed:
Update hostCommunication test

sfc-gh-jgarcia committed:
Merge branch 'sis-updates' of https://github.com/streamlit/streamlit into sis-updates

mayagbarnes commented:
Looks like only remaining test failure is in `App.tsx` for the `"does nothing if the custom theme received from server has a matching hash"` test -> the hash comparison is failing, so `addThemes` and `setTheme` are called. 
This is a result of the `customThemeConfig` constructor initializing a default empty array for the `fontFaces` key, while the NewSession constructor doesn't include an empty `fontFaces` key at all. This is a protobuf issue and Ken/I don't believe it will be problematic, so just updating the test should be okay.
In `NEW_SESSION_JSON` object, resolve per below:
```javascript
  const NEW_SESSION_JSON = {
    config: {
      gatherUsageStats: false,
      maxCachedMessageAge: 0,
      mapboxToken: "mapboxToken",
      allowRunOnSave: false,
      hideSidebarNav: false,
    },
    customTheme: {
      primaryColor: "red",
      fontFaces: [],
    },
// .... rest of the stuff here
```

sfc-gh-jgarcia committed:
Fix App.tsx test

sfc-gh-jgarcia commented:
> c3e6026

Thanks a lot for looking into this, @mayagbarnes! This test had been fixed on [c3e6026](https://github.com/streamlit/streamlit/pull/6393/commits/c3e6026d95eec60c8852c3d0bf0dd70d6d1debf1).

Let me know if you need anything else here!

mayagbarnes commented on frontend/src/hocs/withHostCommunication/types.ts:
```diff
@@ -108,6 +108,10 @@ export type IHostToGuestMessage = {
     }
+  | {
+      type: "SET_CUSTOM_THEME_CONFIG"
+      theme: ICustomThemeConfig
```
Not sure why this doesn't trigger an error, but need to update this payload to be called `themeInfo` consistent with every other place updated in [this commit](https://github.com/streamlit/streamlit/pull/6393/commits/863d59f9d815cd746a9c375fae833278bda1900b)

sfc-gh-jgarcia replied:
Fixed in [0486457](https://github.com/streamlit/streamlit/pull/6393/commits/0486457167675fc033f22a58ba70795f6a7c1a7a)!

mayagbarnes commented on frontend/src/hocs/withHostCommunication/withHostCommunication.tsx:
```diff
@@ -123,6 +129,11 @@ function withHostCommunication<P extends InjectedProps>(
 
       const { allowedOrigins, useExternalAuthToken } = allowedOriginsResp
+
+      if (IS_DEV_ENV) {
+        allowedOrigins.push("http://localhost:8000")
+      }
```
@vdonato: This was added to make manual testing easier using the `frontend/hostframe.html` file & instructions - just want to confirm it isn't problematic

sfc-gh-jgarcia replied:
> Also when I test locally and hard refresh (browser refresh button vs just hitting R to re-run), the theme reverts back to our standard theme vs restoring the theme sent with postMessage. This is intended behavior/unproblematic for SIS right?

@sfc-gh-tteixeira or someone on the SiS team might be able to confirm with a bit more detail than me, but my understanding was that the way SiS infrastructure works, the app will be reloaded every time you access it, and we don't have any caching mechanism for themes there, so the message will always be re-sent, causing the app to update the theme accordingly (tough leading to other potential problems, such as flickering when the theme updates).

vdonato replied:
If possible, I think I'd prefer either

1. Just getting rid of this entirely and temporarily re-adding it when necessary for manual testing, which hopefully won't be so often that this is too annoying
2. Doing the same thing on the backend in the `AllowedMessageOriginsHandler` handler when the `global.developmentMode` config option is true (note that this is `False` in built wheel files).

I don't think it's problematic even having to leave this as-is, but it'd be worth checking with @sfc-gh-hpathak on this.

sfc-gh-tteixeira replied:
I'll just comment it out for now, then. That helps unblock this PR and allows us to have this discussion later.

mayagbarnes reviewed (approved):
Please see comment to fix `withHostCommunication/types.ts` discrepancy, as well as clarification below on intended behavior, but otherwise LGTM 👍🏼 

Will ask Vincent for protobuf codeowner review at standup tomorrow

mayagbarnes commented:
Also when I test locally and hard refresh (browser refresh button vs just hitting R to re-run), the theme reverts back to our standard theme vs restoring the theme sent with postMessage. This is intended behavior/unproblematic for SIS right?

sfc-gh-jgarcia committed:
Update payload for themeInfo in hostCommunication

sfc-gh-jgarcia commented:
> Also when I test locally and hard refresh (browser refresh button vs just hitting R to re-run), the theme reverts back to our standard theme vs restoring the theme sent with postMessage. This is intended behavior/unproblematic for SIS right?

@sfc-gh-tteixeira or someone on the SiS team might be able to confirm with a bit more detail than me, but my understanding was that the way SiS infrastructure works, the app will be reloaded every time you access it, and we don't have any caching mechanism for themes there, so the message will always be re-sent, causing the app to update the theme accordingly (tough leading to other potential problems, such as flickering when the theme updates).

sfc-gh-tteixeira commented:
> Also when I test locally and hard refresh (browser refresh button vs just hitting R to re-run), the theme reverts back to our standard theme vs restoring the theme sent with postMessage. This is intended behavior/unproblematic for SIS right?

The localStorage cache is not required for SiS. It would be great if it worked, but also OK if it doesn't. (And even if it worked, there are questions about when should we clear it. Do we do something smart there today, clearing based on the hash of the config or something?)

vdonato commented on proto/streamlit/proto/NewSession.proto:
```diff
@@ -121,6 +124,33 @@ message CustomThemeConfig {
   string text_color = 4;
   FontFamily font = 5;
   BaseTheme base = 6;
+  string widget_background_color = 7;
+  string widget_border_color = 8;
+  Radii radii = 9;
+  string body_font = 13;
+  string code_font = 14;
+  repeated FontFace font_faces = 15;
+  FontSizes font_sizes = 16;
+}
+
+message FontFace {
+  string url = 1;
+  string family = 2;
+  int32 weight = 3;
+  string style = 4;
+}
+
+message Radii {
+  // In pixels.
+  int32 baseWidgetRadius = 1;
+  int32 checkboxRadius = 2;
+}
+
+message FontSizes {
+  // In pixels.
+  int32 tinyFontSize = 1;
+  int32 smallFontSize = 2;
+  int32 baseFontSize = 3;
```
We'll need to give @sfc-gh-wschmitt a heads up that a bunch of fields are being added here, but I'm assuming this is okay because we probably just nuke the theme `custom_theme` field entirely when sanitizing forward messages

vdonato commented:
> Do we do something smart there today, clearing based on the hash of the config or something?

Ah, we do actually reapply the theme when the hash of the theming config received from the server changes (see the `processThemeInput` function in `App.tsx`), and this is probably what's breaking local theme caching with what we receive from `postMessage` since the server-defined theme will always be different.

vdonato reviewed (approved):
Proto changes LGTM

I also did a quick skim of the code itself and it looked fine to me, but I'm mostly relying on @mayagbarnes' review for an actual code review

sfc-gh-tteixeira committed:
Remove localhost allowedOrigin

mayagbarnes committed:
Resolve lint & update test instructions

sfc-gh-tteixeira commented:
Ok, I'm seeing checkmarks next to both your names in Github, and looking at the discussion it feels like everything is resolved, so I'll just go ahead and merge.

Some future work:
* Figure out caching. See [this message from Vincent](https://github.com/streamlit/streamlit/pull/6393#issuecomment-1492379877)
* Figure out a good behavior for when the theme is set by the user in config.toml
* See if flicker from this solution is bad in practice. If so, implement solutions for that!

sfc-gh-kbregula commented on frontend/src/theme/utils.ts:
```diff
@@ -412,10 +478,12 @@ export const createEmotionTheme = (
-        headingFont: parsedFont,
+        bodyFont: themeInput.bodyFont ? themeInput.bodyFont : parsedFont,
+        headingFont: themeInput.bodyFont ? themeInput.bodyFont : parsedFont,
+        codeFont: themeInput.codeFont ? themeInput.codeFont : parsedFont,
```
@sfc-gh-tteixeira I am not sure if we want the same font in 3 places. This probably caused the regressions. See: 
https://github.com/streamlit/streamlit/issues/6484
